File: scripts/find_missing_castings.py
#!/usr/bin/env python3
"""
find_missing_castings.py
------------------------
Находит casting-страницы Hot Wheels Wiki, которых нет в твоей базе.

Идея: все каст-страницы транслюдят Template:Casting. Спрашиваем у MediaWiki API
полный список таких страниц (list=embeddedin) -> это ПОЛНЫЙ перечень кастов на вики.
Дальше диффим с твоими lnk -> остаются недостающие (премиум-онли осядут тут же).

Запускать ЛОКАЛЬНО (нужен доступ к hotwheels.fandom.com):
    pip install requests
    python3 find_missing_castings.py path/to/your_db.json

Вход  : твой JSON — список объектов, у каждого есть ключ "lnk".
Выход : missing_castings.json — [{"title": "...", "slug": "..."}, ...]
        эти slug скармливаешь СВОЕМУ существующему парсеру каст-страниц.
"""

import json
import sys
import time
import urllib.parse

import requests

API = "https://hotwheels.fandom.com/api.php"
# поставь сюда реальный контакт — Fandom не любит безымянных ботов
HEADERS = {"User-Agent": "HW-collection-sync/1.0 (personal research; you@example.com)"}
SLEEP = 0.5  # пауза между запросами, чтобы не словить бан


def norm(title: str) -> str:
    """Нормализуем ключ каста для сравнения: %xx-декод, _<->пробел, регистр."""
    t = urllib.parse.unquote(str(title))   # %27 -> ' ; %26 -> & и т.д.
    t = t.replace("_", " ").strip()
    return t.casefold()


def to_slug(title: str) -> str:
    """Титул страницы -> slug в формате твоего lnk (с подчёркиваниями)."""
    return title.replace(" ", "_")


def all_casting_titles() -> list[str]:
    """Полный список страниц (ns=0), транслюдящих Template:Casting."""
    titles, cont = [], {}
    while True:
        params = {
            "action": "query",
            "list": "embeddedin",
            "eititle": "Template:Casting",
            "einamespace": "0",        # только основные статьи
            "eilimit": "500",
            "format": "json",
            **cont,
        }
        r = requests.get(API, params=params, headers=HEADERS, timeout=30)
        r.raise_for_status()
        data = r.json()
        titles.extend(m["title"] for m in data["query"]["embeddedin"])
        if "continue" in data:
            cont = data["continue"]
            time.sleep(SLEEP)
        else:
            break
    return titles


def resolve_redirects(titles: list[str]) -> dict[str, str]:
    """Схлопываем редиректы: alias -> настоящий титул (чтобы не дублировать)."""
    mapping = {}
    for i in range(0, len(titles), 50):
        chunk = titles[i:i + 50]
        params = {
            "action": "query",
            "redirects": "1",
            "titles": "|".join(chunk),
            "format": "json",
        }
        r = requests.get(API, params=params, headers=HEADERS, timeout=30)
        r.raise_for_status()
        for red in r.json().get("query", {}).get("redirects", []):
            mapping[red["from"]] = red["to"]
        time.sleep(SLEEP)
    return mapping


def load_existing(db_path: str) -> set[str]:
    with open(db_path, encoding="utf-8") as f:
        db = json.load(f)
    keys = set()
    for rec in db:
        lnk = rec.get("lnk")
        if lnk:
            keys.add(norm(lnk))
    return keys


def main():
    db_path = sys.argv[1] if len(sys.argv) > 1 else "your_db.json"

    print(f"[*] читаю твою базу: {db_path}")
    existing = load_existing(db_path)
    print(f"    кастов в базе: {len(existing)}")

    print("[*] тяну полный список каст-страниц с вики (Template:Casting)...")
    wiki_titles = all_casting_titles()
    print(f"    всего каст-страниц на вики: {len(wiki_titles)}")

    # схлопываем редиректы, чтобы alias не считался отдельным кастом
    redir = resolve_redirects(wiki_titles)
    canonical = []
    seen = set()
    for t in wiki_titles:
        real = redir.get(t, t)
        if norm(real) not in seen:
            seen.add(norm(real))
            canonical.append(real)

    missing = [t for t in canonical if norm(t) not in existing]
    missing.sort()

    out = [{"title": t, "slug": to_slug(t)} for t in missing]
    with open("missing_castings.json", "w", encoding="utf-8") as f:
        json.dump(out, f, ensure_ascii=False, indent=2)

    print(f"[+] недостающих кастов: {len(missing)}")
    print(f"[+] записал -> missing_castings.json")
    print("\nпервые 20:")
    for t in missing[:20]:
        print("   ", t)


if __name__ == "__main__":
    main()

File: scripts/sync/find-missing-castings.js
#!/usr/bin/env node
/*
 * find-missing-castings.js
 * ------------------------
 * Finds Hot Wheels Wiki casting pages that are NOT in our database.
 *
 * Every casting page transcludes Template:Casting. We ask the MediaWiki API for
 * the full list of such pages (list=embeddedin) — that is the complete set of
 * castings on the wiki — then diff it against the `lnk` keys in our DB. What is
 * left over is the set of castings we are missing (premium-only ones included).
 *
 * Node port of the original Python script (Python is not installed locally and
 * Node is the project runtime). Uses the global fetch from Node 18+, no deps.
 *
 * Usage:
 *   node scripts/find-missing-castings.js [path/to/db.json]
 *
 * Input  : our JSON — an array of objects, each with a "lnk" key.
 * Output : scripts/output/missing_castings.json — [{ title, slug }, ...]
 *          feed those slugs into the existing casting-page parser.
 */

const fs = require('fs');
const path = require('path');

const API = 'https://hotwheels.fandom.com/api.php';
// Fandom dislikes anonymous bots — identify with a real contact.
const HEADERS = { 'User-Agent': 'HWDB-collection-sync/1.0 (personal research; oreanor@gmail.com)' };
const SLEEP_MS = 500; // pause between requests to stay polite

const sleep = (ms) => new Promise((resolve) => setTimeout(resolve, ms));

// Normalize a casting key for comparison: %xx-decode, _<->space, casefold.
function norm(title) {
  let t;
  try {
    t = decodeURIComponent(String(title));
  } catch {
    t = String(title);
  }
  return t.replace(/_/g, ' ').trim().toLowerCase();
}

// Page title -> slug in our `lnk` format (underscores instead of spaces).
const toSlug = (title) => title.replace(/ /g, '_');

async function apiGet(params) {
  const url = `${API}?${new URLSearchParams(params).toString()}`;
  const res = await fetch(url, { headers: HEADERS });
  if (!res.ok) throw new Error(`API ${res.status} for ${url}`);
  return res.json();
}

// Full list of ns=0 pages transcluding Template:Casting.
async function allCastingTitles() {
  const titles = [];
  let cont = {};
  for (;;) {
    const data = await apiGet({
      action: 'query',
      list: 'embeddedin',
      eititle: 'Template:Casting',
      einamespace: '0',
      eilimit: '500',
      format: 'json',
      ...cont,
    });
    for (const m of data.query.embeddedin) titles.push(m.title);
    if (data.continue) {
      cont = data.continue;
      process.stdout.write(`\r    fetched ${titles.length} titles...`);
      await sleep(SLEEP_MS);
    } else {
      break;
    }
  }
  process.stdout.write('\n');
  return titles;
}

// Collapse redirects: alias -> real title (so an alias is not a separate casting).
async function resolveRedirects(titles) {
  const mapping = {};
  for (let i = 0; i < titles.length; i += 50) {
    const chunk = titles.slice(i, i + 50);
    const data = await apiGet({
      action: 'query',
      redirects: '1',
      titles: chunk.join('|'),
      format: 'json',
    });
    for (const red of data.query?.redirects ?? []) mapping[red.from] = red.to;
    process.stdout.write(`\r    resolved ${Math.min(i + 50, titles.length)}/${titles.length} redirects...`);
    await sleep(SLEEP_MS);
  }
  process.stdout.write('\n');
  return mapping;
}

function loadExisting(dbPath) {
  const db = JSON.parse(fs.readFileSync(dbPath, 'utf-8'));
  const keys = new Set();
  for (const rec of db) {
    if (rec.lnk) keys.add(norm(rec.lnk));
  }
  return keys;
}

async function main() {
  const dbPath = process.argv[2] || path.join('data', 'hw.json');

  console.log(`[*] reading our DB: ${dbPath}`);
  const existing = loadExisting(dbPath);
  console.log(`    castings in DB: ${existing.size}`);

  console.log('[*] pulling the full casting list from the wiki (Template:Casting)...');
  const wikiTitles = await allCastingTitles();
  console.log(`    casting pages on the wiki: ${wikiTitles.length}`);

  console.log('[*] resolving redirects...');
  const redir = await resolveRedirects(wikiTitles);
  const canonical = [];
  const seen = new Set();
  for (const t of wikiTitles) {
    const real = redir[t] ?? t;
    if (!seen.has(norm(real))) {
      seen.add(norm(real));
      canonical.push(real);
    }
  }

  const missing = canonical.filter((t) => !existing.has(norm(t))).sort();
  const out = missing.map((t) => ({ title: t, slug: toSlug(t) }));

  const outDir = path.join('scripts', 'output');
  fs.mkdirSync(outDir, { recursive: true });
  const outPath = path.join(outDir, 'missing_castings.json');
  fs.writeFileSync(outPath, JSON.stringify(out, null, 2), 'utf-8');

  console.log(`[+] missing castings: ${missing.length}`);
  console.log(`[+] wrote -> ${outPath}`);
  console.log('\nfirst 20:');
  for (const t of missing.slice(0, 20)) console.log('   ', t);
}

main().catch((err) => {
  console.error('Failed:', err);
  process.exit(1);
});
